substrate/primitives/genesis-builder/src/lib.rs

```diff
@@ -19,39 +19,92 @@
-//!   pallet's author.
-//! - deserialize the `RuntimeGenesisConfig` from given json blob and put `RuntimeGenesisConfig`
-//!   into the state storage. Allows to build customized configuration.
+//! This module provides means to interact with `RuntimeGenesisConfig`. Runtime provides a default
+//! `RuntimeGenesisConfig` structure in a form of the JSON blob.
 //!
-//! Providing externalities with empty storage and putting `RuntimeGenesisConfig` into storage
-//! allows to catch and build the raw storage of `RuntimeGenesisConfig` which is the foundation for
-//! genesis block.
+//! Additionally the runtime may provide a number of partial predefined `RuntimeGenesisConfig`
+//! configurations in the form of patches which shall be applied on top of the default
+//! `RuntimeGenesisConfig`. The patch is a JSON blob, which essentially comprises the list of
+//! key-value pairs that are to be customized in the default runtime genesis config.
+//! These predefined configurations are referred to as presets.
+//!
+//! This allows the runtime to provide a number of predefined configs (e.g. for different
+//! testnets or development) without neccessity to leak the runtime types outside the itself (e.g.
+//! node or chain-spec related tools).
+//!
+//! This Runtime API allows to interact with `RuntimeGenesisConfig`, in particular:
+//! - provide the list of available preset names,
+//! - provide a number of named presets of `RuntimeGenesisConfig`,
+//! - provide a JSON represention of the default `RuntimeGenesisConfig` (by simply serializing the
+//!   default `RuntimeGenesisConfig` struct into JSON format),
+//! - deserialize the full `RuntimeGenesisConfig` from given JSON blob and put the resulting
+//!   `RuntimeGenesisConfig` structure into the state storage creating the initial runtime's state.
+//!   Allows to build customized genesis. This operation internally calls `GenesisBuild::build`
+//!   function for all runtime pallets.
+//!
+//! Providing externalities with an empty storage and putting `RuntimeGenesisConfig` into storage
+//! (by calling `build_state`) allows to construct the raw storage of `RuntimeGenesisConfig`
+//! which is the foundation for genesis block.
 
 extern crate alloc;
+use alloc::vec::Vec;
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
 
 /// The result type alias, used in build methods. `Err` contains formatted error message.
 pub type Result = core::result::Result<(), sp_runtime::RuntimeString>;
 
+/// Wrapper representing preset id.
+#[derive(Decode, Encode, Debug, TypeInfo, PartialEq)]
+pub struct PresetId(Vec<u8>);
+
+impl From<&str> for PresetId {
+	fn from(v: &str) -> Self {
+		Self(v.as_bytes().to_vec())
+	}
+}
+
+impl<'a> TryFrom<&'a PresetId> for &'a str {
+	type Error = core::str::Utf8Error;
+	fn try_from(v: &'a PresetId) -> core::result::Result<&'a str, Self::Error> {
+		core::str::from_utf8(&v.0)
+	}
+}
+
 sp_api::decl_runtime_apis! {
```

Comment:
Does this resonate with you as a good representation of the user flow? 

https://mermaid.live/edit#pako:eNqFUU1LxDAU_Cvlndsl_di0zcGLiiB6qTeNLLF97QbapCQp7lr63822InhYfKc3w8w8HjNDrRsEBm2vP-ujMC54qrgK_NyfHBol-g09oEIr7a1WrewerVYbXU3KyQEDrv6agii6-c8TRR26w2jQorsiv3LbWz8m2TcH64TDi7dK3n5i37mCEAY0g5CNf2y-ZHBwRxyQA_Nrg62YeseBq8VLxeT0y1nVwFrRWwxhGhufeidFZ8Twy45CvWrtsTPTBoHNcAJWpjualHtS0iQr85QUIZyBxXmyK2hapgXJaZESki0hfK0BZNUnNE49v6c0i0PARjptnrcu1kqWbz59g20

Reply:
the narration would be: You always need a genesis config json file. You can either get this from the runtime by generating it from a preset (so the outcome of a preset is a genesis config), or from other extranl sources, like `chain-spec-builder`.

Reply:
Yes.

But my mental model is:
-  _external_ is somehow hand-crafted or edited/patched genesis config json file (e.g. containing manually added endowed account or parachain configuration for zombie network).
- `chain-spec-builder` is just an util that helps us to build the chain-spec using different source, one being runtime's `GenesisBuilder`.